integration/dynamic_namespace_delete_test.go

```diff
@@ -0,0 +1,176 @@
+	"github.com/m3db/m3x/instrument"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/stretchr/testify/require"
+	"github.com/uber-go/tally"
+)
+
+func TestDynamicNamespaceDelete(t *testing.T) {
+	if testing.Short() {
+		t.SkipNow() // Just skip if we're doing a short run
+	}
+
+	// test options
+	testOpts := newTestOptions(t)
+	require.True(t, len(testOpts.Namespaces()) >= 2)
+	ns0 := testOpts.Namespaces()[0]
+	ns1 := testOpts.Namespaces()[1]
+
+	reporter := xmetrics.NewTestStatsReporter(xmetrics.NewTestStatsReporterOptions())
+	scope, closer := tally.NewRootScope(tally.ScopeOptions{Reporter: reporter}, 100*time.Millisecond)
+	defer closer.Close()
+
+	// embedded kv
+	embeddedKV, err := etcd.New(etcd.NewOptions())
+	require.NoError(t, err)
+	defer func() {
+		require.NoError(t, embeddedKV.Close())
+	}()
+	require.NoError(t, embeddedKV.Start())
+	csClient, err := embeddedKV.ConfigServiceClient()
+	require.NoError(t, err)
+	kvStore, err := csClient.KV()
+	require.NoError(t, err)
+
+	// namespace maps
+	protoKey := func(nses ...namespace.Metadata) proto.Message {
+		nsMap, err := namespace.NewMap(nses)
+		require.NoError(t, err)
+		return convert.ToProto(nsMap)
+	}
+
+	// dynamic namespace registry options
+	dynamicOpts := dynamic.NewOptions().
+		SetConfigServiceClient(csClient).
+		SetInstrumentOptions(instrument.NewOptions().SetMetricsScope(scope))
+	dynamicInit := dynamic.NewInitializer(dynamicOpts)
+	testOpts = testOpts.SetNamespaceInitializer(dynamicInit)
+
+	// initialize value in kv
+	_, err = kvStore.Set(dynamicOpts.NamespaceRegistryKey(), protoKey(ns1))
+	require.NoError(t, err)
+
+	// Test setup
+	testSetup, err := newTestSetup(t, testOpts)
+	require.NoError(t, err)
+	defer testSetup.close()
+
+	// Start the server
+	log := testSetup.storageOpts.InstrumentOptions().Logger()
+	require.NoError(t, testSetup.startServer())
+
+	// Stop the server
+	defer func() {
+		require.NoError(t, testSetup.stopServer())
+		log.Infof("server is now down")
+	}()
+
+	// Write test data
+	blockSize := ns0.Options().RetentionOptions().BlockSize()
+	now := testSetup.getNowFn()
+	seriesMaps := make(map[time.Time]generate.SeriesBlock)
+	inputData := []generate.BlockConfig{
+		{[]string{"foo", "bar"}, 100, now},
+		{[]string{"foo", "baz"}, 50, now.Add(blockSize)},
+	}
+	for _, input := range inputData {
+		start := input.Start
+		testData := generate.Block(input)
+		seriesMaps[start] = testData
+	}
+	log.Infof("test data is now generated")
+
+	// fail to write to non-existent namespaces
+	for _, testData := range seriesMaps {
+		require.Error(t, testSetup.writeBatch(ns0.ID(), testData))
+	}
+
+	// delete namespace key, ensure update propagates
+	numInvalid := numInvalidNamespaceUpdates(scope)
+	_, err = kvStore.Delete(dynamicOpts.NamespaceRegistryKey())
+	require.NoError(t, err)
+	log.Infof("deleted namespace key from kv")
+	for {
+		invalidUpdates := numInvalidNamespaceUpdates(scope)
+		if invalidUpdates > numInvalid {
+			break
+		}
```

Comment:
nit: Can rewrite this as `for numInvalidNamespaceUpdates(scope) <= numInvalid { time.Sleep(time.Second) }`

Reply:
sure, will just make it use `waitUntil` to follow the pattern below

Reply:
done